### DISCUSSION TITLE
외부 스위치를 이용해 월을 바꿀 수 있는 방법이 없을까요?

### DISCUSSION THREAD

**Discussion by PaekHyun** [MMRIZE/MMM-CalendarExt3#168] [Q&A]
외부에 설치된 버튼 (스위치?)를 통해 
월을 바꿀 수 있는 방법이 있는지 궁금합니다.

chatgpt말로는

const Gpio = require('onoff').Gpio; // onoff 모듈로 GPIO 제어
const buttonNext = new Gpio(18, 'in', 'both'); // 다음달로 이동할 버튼
const buttonPrev = new Gpio(23, 'in', 'both'); // 지난달로 이동할 버튼

Module.register("MMM-CalendarExt3", {
    start: function() {
        this.sendSocketNotification("INIT");
        this.setupButtonListeners();  // GPIO 버튼 핸들러 설정
    },

    setupButtonListeners: function() {
        // 버튼 눌림 감지 핸들러
        buttonNext.watch((err, value) => {
            if (value === 1) {
                this.sendNotification("SHOW_NEXT_MONTH");
            }
        });

        buttonPrev.watch((err, value) => {
            if (value === 1) {
                this.sendNotification("SHOW_PREVIOUS_MONTH");
            }
        });
    },

    notificationReceived: function(notification, payload, sender) {
        if (notification === "SHOW_NEXT_MONTH") {
            // 다음 달로 변경하는 로직
            this.changeMonth(1);  // 1은 다음 달로 이동을 의미
        }

        if (notification === "SHOW_PREVIOUS_MONTH") {
            // 지난 달로 변경하는 로직
            this.changeMonth(-1);  // -1은 지난 달로 이동을 의미
        }
    },

    changeMonth: function(direction) {
        // 달력 월 변경 로직 구현 (예: CalendarExt3 모듈의 메소드 호출)
        this.sendSocketNotification("CHANGE_MONTH", { direction: direction });
    }
});

MMM-CalendarExt3.js 파일에 저렇게 넣으라고 조언을 해줘서..

했지만 결론적으로 저기 저 onoff모듈이 오래된건지 무슨 문제가 있는지 첫줄부터 안되더라구요.
const Gpio = require('onoff').Gpio; // onoff 모듈로 GPIO 제어
const buttonNext = new Gpio(18, 'in', 'both'); // 다음달로 이동할 버튼
const buttonPrev = new Gpio(23, 'in', 'both'); // 지난달로 이동할 버튼

조언 부탁드리기 위해 글 남겨봅니다.

감사합니다.

**Comment by eouia** [comment]
ChatGPT가 짜준 코드는... 뭐.. 전혀 엉망진창이구요. ^^;;; 그럴 듯 해보이긴 하는데 첫단추부터 잘못 되었습니다.

일단 MagicMirror 모듈들의 아키텍쳐는 기본적으로;
모듈들간의 통신은 `notification`이라는 것을 사용해요.
A라는 모듈에서 B라는 모듈로 뭔가 신호를 보내고 싶다면, 
`A.sendNotification("SOMETHING", payload)` 처럼 notification을 보내면
`B.notificationReceived(notification, payload, sender)`라는 callback 함수가 호출되어 해당 notification과 payload를 처리하는 식으로 되어 있습니다.
이 때, B 모듈은 자기가 처리해야 할 notification들을 취사 선택해서 처리하기 때문에 아무 notification이나 보내도 되는 건 아니고, B 모듈이 준비되어 있고, 처리가능한 notification만 유효하지요.

현재 의도하고 계신 기능은 두가지로 나눠서 생각해봐야 하는데요;
1) GPIO를 사용하는 H/W 버튼을 눌렀을 때 특정 notification을 발생시킨다.
2) 해당 notification을 받아서 CX3의 뷰를 갱신한다.

우선 2)에 대해서;
MMM-CalendarExt3모듈에서는 `CX3_SET_CONFIG` 이라는 notification을 받으면 해당 payload로 현재 모듈의 config을 대체하도록 준비되어 있습니다. (https://github.com/MMRIZE/MMM-CalendarExt3?tab=readme-ov-file#notification)

어떤 모듈이 
```
"CX3_SET_CONFIG", { monthIndex: 1 }"
``` 
이라는 notification을 보낸다면, CX3는 그 즉시 현재 기준 다음달 뷰를 보여주는 식입니다.
(현재 기준이니까 버튼을 연달아 눌러도 무조건 "다음달" 만 보여주긴 합니다. 이걸 누적 적용되어 계속 `그 다음달, 그 다음달`처럼 작동려면 좀 더 복잡해지긴 합니다만, https://github.com/MMRIZE/MMM-CalendarExt3?tab=readme-ov-file#cx3_reset-payload--callback-instanceid- 이게 그에 대한 샘플 코드입니다.)

여하튼, CX3는 receiver니까 다른 어떤 모듈이 해당 notification을 shoot 해주어야겠지요? CX3에 H/W 버튼 제어기능을 직접적으로 넣는 것은 MagicMirror 모듈 설계 개념에 어긋나니까, 이 역할은 다른 모듈이 맡아야겠지요.

> 물론, 모든 것을 한 모듈안에서 다 처리하는 거대 모듈을 작성할 수도 있습니다만,
> - 아무리 거대 모듈이래도 사용자의 요구는 끝이 없고,
> - 유지보수가 쓸 데 없이 복잡해지며,
> - 적절하게 역할 분담된 모듈들의 조합으로 더 많은 기능을 제공할 수 있으므로
> 가능한 작은 모듈들을 유지하는 게 더 좋습니다.

그래서 1) 의 내용으로 들어가자면;
이미 GPIO로 연결된 버튼의 입력신호를 받아 특정 notification으로 변환해주는 모듈들이 있습니다. 꽤 많이요.
예를 들자면, https://github.com/MarcLandis/MMM-Buttons 같은 모듈이 적절해 보이네요. (사용은 안해봤습니다만. ^^;)
이 모듈을 사용하면 버튼이 눌리면 미리 정의된 notification을 발신할 수 있으니, 그 notification과 payload로 CX3 모듈이 알아들을 수 있는, 필요한 notification을 보낼 수 있도록 정의하시면 되겠죠.


물론, 기존 모듈 대신 직접 만드셔도 됩니다. 다만 ChatGPT가 짜준 코드는 그럴 듯 해보이긴 한데 틀린 코드이고 (새 모듈을 제작하는 게 아니라, CX3를 재정의 하려고 하시네요. 당연히 제대로 동작하지 않습니다.), 모듈 작성법은 재미있는 주제이긴 한데, 이 쓰레드안에서 논의할 내용은 아닌 것 같아 생략합니다.

**Comment by PaekHyun** [comment]
먼저 답변 감사합니다.

저는 [MMM-GPIO-Notification](https://github.com/Tom-Hirschberger/MMM-GPIO-Notifications)
이 모듈을 설치해서 nofitication을 날려서 다음 달(payload를 수정하니 이전달도...)을 볼 수 있게 되었습니다.

{
			module: "MMM-GPIO-Notifications",
			config: {			
				'17': { // GPIO 핀 번호 (17번 예시)					
					notifications_low : [
						{
							notification : 'CX3_SET_CONFIG', 
							payload: {'monthIndex' : 1},
						}	
					]
				},
			}
		},
정말 간단하게 구현이 되더라구요..

그런데 궁극적으로는 버튼 두개를 외부에 붙여서 달을 변경하고 싶은 것이 목적입니다.
그래서 두번째 말씀해주신 그 소스를 봤는데.


this.sendNotification('CX3_GET_CONFIG', {  
  callback: (current) => {
    console.log(current.mode, current.monthIndex)
  }
})

this.sendNotification('CX3_GET_CONFIG', {
  callback: (before) => {
    this.sendNotification('CX3_SET_CONFIG', {
      monthIndex: before.monthIndex + 1 })
  }
})

이 함수 그대로 넣으면 될 것 같은데.... 
이걸 어디에 넣어야 할까요??? MMM-CalendarExt3.js 파일 어딘가에 넣는걸까요?
(CS_RESET 이 부분은 아마도 10초후에 원래대로 돌린다는 의미인것 같아서 삭제..했습니다.)

그리고 만약에 저대로 한다면..
payload를 아무것도 쓰지 않아도 되는것 같은데..맞을까요??

**Comment by eouia** [comment]
```javascript
.sendNotification('SOMETHING', payload)
```
에서,  `SOMETHING`이 notification이고, `payload` 변수로 전달되는 값이 payload object에요. 어떤 값이 들어와도 상관없죠.
그러니까, 
```javascript
{
  callback: (before) => {
    this.sendNotification('CX3_SET_CONFIG', { monthIndex: before.monthIndex + 1 })
  }
}
```
이것 자체가 전달할 payload object에요. 

`MMM-GPIO-notification`에서 사용하시려면,
```javascript
{
  module: 'MMM-GPIO-Notifications',
  config: {
    '17': {
      notifications_high: [
        {
          notification: 'CX3_GET_CONFIG',
          payload: {
            callback: (before) => {
              this.sendNotification('CX3_SET_CONFIG', {
                monthIndex: before.monthIndex + 1
              })
            }
          }
        },
      ]
    },
  }
},
```
아마 이런 식으로 써야 할 텐데요... 그런데 문제가 있네요.

`MMM-GPIO-Notification`을 들여다봤는데요, 공교롭게도, 해당 모듈은 payload를 socket.io를 통해 node_helper로 보내기 위해 JSONify 해버려서, payload로 function등을 다이렉트로 전달할 수 없는 구조네요. (왜 그렇게 만들었을까...) 단순한 `{ monthIndex: 1 }` 같은 값은 `payload`로 전달할 수 있지만, `{ callback: () => {...} }`같은 펑션을 전달하려고 하면 중간에 사라져버려서 `payload`로 전달할 수가 없어요.

가능한 대안은,

1) 다른 모듈을 사용하던가,
2) MMM-GPIO-Notifications 제작자에게 function 등을 payload로 실어보낼 수 있게 고쳐달라고 하던가,
3) MMM-GPIO-Notifications 만으로는 우리가 필요한 function을 payload로 전달할 수 없으므로, MMM-GPIO-Notifications로는 단순한 notification을 발하고, 그걸 받아서 우리가 필요로 하는 notification으로 변환해주는 모듈을 추가로 사용하는 방법이 있겠죠. `MMM-NotificationTrigger`를 쓰면 될 것 같긴 합니다만... 

세가지 다 그냥 이렇게 하세요 식으로 간단히 설명드리기는 좀 어렵네요. ㅎㅎㅎ 일단 1)을 추천드리고, 만약 안되면 3)에 대해서 별도로 설명드리도록 할게요.

**Comment by PaekHyun** [comment]
안녕하세요 답변 감사드립니다.

세가지 대안을 찾아봤는데.

1. 못찾겠습니다. ㅠㅠ 몇가지 key 입력 등 여기저기 해봐도 대부분 json으로 입력 받는것 같더라구요.. 그래서 ...안되는것같고.
2. 될지 안될지도 모르고 해준다고 해도 오래 걸릴것 같고..........해서.
3. 3번으로 해봤습니다...
그런데..

먼저 테스트 삼아 아래 코드를 넣어봤습니다.
```

{
	module : 'MMM-GPIO-Notifications', 
	config : {
		'17' : {
			notifications_low : [
				{ 
					notification : 'NEXT_MONTH', 
				}
			]			
		}
	}
},
{
  module: 'MMM-NotificationTrigger',  
  config: {
    useWebhook: false, 
    triggers:[ 
      {
        trigger: "NEXT_MONTH", 
        fires: [ 
          {
            fire:"CX3_SET_CONFIG", //REQUIRED
            payload : { 'monthIndex' : 1 },
            },         
          },
        ],
      },
    ]
  },
},
```


이렇게 하니 정상적으로 다음 달 표시가 나오더라구요..
그래서 말씀하신 함수를 아래와 같이 넣어 봤는데..

```

{
	module : 'MMM-GPIO-Notifications', 
	config : {
		'17' : {
			notifications_low : [
				{ 
					notification : 'NEXT_MONTH', 
					//payload : { 'monthIndex' : 1 } , 
				}
			]			
		}
	}
},
{
  module: 'MMM-NotificationTrigger',  
  config: {
    useWebhook: false, 
    triggers:[ 
      {
        trigger: "NEXT_MONTH", //REQUIRED     
        fires: [ // Array of fires. You can enable multi-firing with one trigger.
          {
            fire:"CX3_GET_CONFIG", //REQUIRED
            payload : { callback: (before) => { this.sendNotification('CX3_SET_CONFIG', { monthIndex: before.monthIndex + 1 }) }      
            },              
          },
        ],
      },
    ]
  },
},

```

아무런 반응이 없습니다..
어떻게 수정을 해야 할까요??

**Reply by eouia** [reply]
`MMM-NotificationTrigger` config을 이렇게 해보세요.
```javascript
{
	module: "MMM-NotificationTrigger",
	config: {
		useWebhook: false,
		triggers: [
			{
				trigger: "NEXT_MONTH",
				fires: [
					{
						fire: "CX3_GET_CONFIG",
						payload: {
							callback: (currentMonthConfig) => {
								const self = MM.getModules().find(m => m.name === "MMM-NotificationTrigger")
								self.sendNotification("CX3_SET_CONFIG", {
									monthIndex: currentMonthConfig.monthIndex + 1
								})
							}
						}
					}
				]
			},
			{
				trigger: "PREV_MONTH",
				fires: [
					{
						fire: "CX3_GET_CONFIG",
						payload: {
							callback: (currentMonthConfig) => {
								const self = MM.getModules().find(m => m.name === "MMM-NotificationTrigger")
								self.sendNotification("CX3_SET_CONFIG", {
									monthIndex: currentMonthConfig.monthIndex - 1
								})
							}
						}
					}
				]
			}
		]
	}
},
```

참조하셨던 예제 샘플은 모듈 안에서 직접 `.sendNotification` 멤버메쏘드를 호출하는 거라서, `this`가 모듈 그 자체를 가리키지만,
MMM-NotificationTrigger에서 config으로 정의할 때에는 `this`가 특정 모듈에 binding되지 않았으므로, this가 windows객체를 가리키게 됩니다. 그래서 `this.sendNotification`이 작동하지 않았어요. 그래서 임의의 모듈을 찾아서(MM.getModules().find()) 그 모듈이 sendNotification한 것처럼 작성합니다.

MMM-NotificationTrigger는 오래 전에 만든 코드라 저런 부분을 감안안했네요. 지금 다시 만들라고 하면 아마 helper 오브젝트 같은 것을 주어서 이런 작업을 편하게 하도록 했을텐데.

**Comment by PaekHyun** [comment]
정말 감사합니다. ㅠㅠ 드디어 잘 되네요 ㅠㅠ

달력 만들기 쉬울 줄 알았는데 정말 쉽지 않네요..
구글링 하면 오래된 코드들이라 제대로 동작을 안하거나..
잘못 알려진 부분이 복붙되서 퍼지면서 안되는 경우도 너무 많고 ...
(예를 들면 센서로 모니터 껐다가 켜는 코드라던가..)

이제 정말 거의 다 된 것 같습니다.

움직임 감지 남은 시간 카운팅해주는 걸 OLED로 보여주는 코드 추가하는거랑
초음파 센서 / 스위치 / OLED 케이스 만들어서 모니터에 붙이기만 하면 될 것 같습니다.

도움 정말 감사드립니다.

**Reply by eouia** [reply]
완성되면 함 보여주세요. ^^

**Reply by PaekHyun** [reply]
알겠습니다. ㅋ
완성하면 한번 사진 올려볼게요..^^

**Comment by PaekHyun** [comment]
감사합니다. 덕분에 잘 만들었어요..
외부 버튼으로 메시지 전달해서 달이 바뀌긴하는데 ..쪼끔 멍청하다고 해야 하나..
한번에 안되기도 하고 여러번 눌러야 하기도 하고..달이 두달 넘어가기도 하고 ;;;;
하지만 그래도 되긴하네요 (어차피 자주 할것도 아니고 ..월말 정도에나 쓸거라..)

초음파 센서 두개로 움직임을 감지해서 화면을 끄는 코드도 넣었는데.
초음파 센서가 한개일 땐 문제없이 동작하는데 두개일 땐 하나가 오동작하는거 외에는..
완성되었습니다. 

마무리 된 다음에 3D프린터로 케이스 작업하고 붙이느라 시간이 좀 더 걸렸네요 

감사합니다 .덕분에 잘 마무리 되었어요..

![1728555958554-0](https://github.com/user-attachments/assets/07ec93f7-d56f-4769-b42a-23bb344debe8)
![1728555958554-1](https://github.com/user-attachments/assets/eae49b3b-f266-4625-9a56-56158a509ca7)
